src/lib/snarky_field_extensions/field_extensions.ml

```diff
@@ -636,3 +682,149 @@ module Cyclotomic_square = struct
         , fpos bsq1 c1 ) )
   end
 end
+
+module F6
+    (Fq : Intf.S with type 'a A.t = 'a and type 'a Base.t_ = 'a)
+    (Fq2 : Intf.S_with_primitive_element
+           with module Impl = Fq.Impl
+            and type 'a A.t = 'a * 'a
+            and type 'a Base.t_ = 'a Fq.t_) (Fq3 : sig
+        include
+          Intf.S_with_primitive_element
+          with module Impl = Fq.Impl
+           and type 'a A.t = 'a * 'a * 'a
+           and type 'a Base.t_ = 'a Fq.t_
+
+        module Params : sig
+          val non_residue : Fq.Unchecked.t
+
+          val frobenius_coeffs_c1 : Fq.Unchecked.t array
+
+          val frobenius_coeffs_c2 : Fq.Unchecked.t array
+        end
+    end) (Params : sig
+      val frobenius_coeffs_c1 : Fq.Unchecked.t array
+    end) =
+struct
+  include E2
+            (Fq3)
+            (struct
+              let non_residue : Fq3.Unchecked.t =
+                Fq.Unchecked.(zero, one, zero)
+
+              let mul_by_non_residue = Fq3.mul_by_primitive_element
+            end)
+
+  let fq_mul_by_non_residue x = Fq.scale x Fq3.Params.non_residue
+
+  (* Number to beat: 3 constraints + 2 fq3 muls *)
+  let special_mul (a0, a1) (b0, b1) =
```

Comment:
any context for this? Libsnark code or papers? and/or a derivation in a comment?

Reply:
whoops -- i just wrote that for myself, i should remove it.